nemo/collections/common/tokenizers/sentencepiece_tokenizer.py

```diff
@@ -0,0 +1,259 @@
+
+            ids.extend(self.tokenizer.encode_as_ids(text[idx:next_idx]))
+            ids.append(self.special_token_to_id[next_token])
+            idx = next_idx + len(next_token)
+
+        ids.extend(self.tokenizer.encode_as_ids(text[idx:]))
+        return ids
+
+    def tokens_to_text(self, tokens):
+        return self.tokenizer.decode_pieces(tokens)
+
+    def ids_to_text(self, ids):
+        text = ""
+        last_i = 0
+
+        for i, id in enumerate(ids):
+            if id in self.id_to_special_token:
+                text += self.tokenizer.decode_ids(ids[last_i:i]) + " "
+                text += self.id_to_special_token[id] + " "
+                last_i = i + 1
+
+        text += self.tokenizer.decode_ids(ids[last_i:])
+        return text.strip()
+
+    def tokens_to_ids(self, tokens):
+        ids = []
+        for token in tokens:
+            ids.append(self.token_to_id(token))
+        return ids
+
+    def token_to_id(self, token):
+        if token in self.special_token_to_id:
+            return self.special_token_to_id[token]
+        return self.tokenizer.piece_to_id(token)
+
+    def ids_to_tokens(self, ids):
+        tokens = []
+        for id in ids:
+            if id >= self.original_vocab_size:
+                tokens.append(self.id_to_special_token[id])
+            else:
+                tokens.append(self.tokenizer.id_to_piece(id))
+        return tokens
+
+    def add_special_tokens(self, special_tokens):
+        if isinstance(special_tokens, list):
+            for token in special_tokens:
+                if (
+                    self.tokenizer.piece_to_id(token) == self.tokenizer.unk_id()
+                    and token not in self.special_token_to_id
+                ):
+                    self.special_token_to_id[token] = self.vocab_size
+                    self.id_to_special_token[self.vocab_size] = token
+                    self.vocab_size += 1
+        elif isinstance(special_tokens, dict):
+            for token_name, token in special_tokens.items():
+                setattr(self, token_name, token)
+                if (
+                    self.tokenizer.piece_to_id(token) == self.tokenizer.unk_id()
+                    and token not in self.special_token_to_id
+                ):
+                    self.special_token_to_id[token] = self.vocab_size
+                    self.id_to_special_token[self.vocab_size] = token
+                    self.vocab_size += 1
+
+    @property
+    def pad_id(self):
+        return self.tokens_to_ids([self.pad_token])[0]
+
+    @property
+    def bos_id(self):
+        return self.tokens_to_ids([self.bos_token])[0]
+
+    @property
+    def eos_id(self):
+        return self.tokens_to_ids([self.eos_token])[0]
+
+    @property
+    def sep_id(self):
+        return self.tokens_to_ids([self.sep_token])[0]
+
+    @property
+    def cls_id(self):
+        return self.tokens_to_ids([self.cls_token])[0]
+
+
+def create_spt_model(
+    data_file: str,
```

Comment:
Can we move this method inside SentencePieceTokenizer and create the model if it doesn't exist?

Reply:
I would not do this, but rather keep create_spt_model, and class SentencePieceTokenizer modular with clean api, and rely on a helper function `get_tokenizer` to check if the model exists. Both create_spt_model and SentencePieceTokenizer kept in the same file for this purpose. i will add  a commentin  SentencePieceTokenizer  which refers to create_spt_model